File: backend/internal/awsx/insights.go
package awsx

// CloudWatch Logs Insights runner, ported from logsinsights.ts. One query with
// a hard deadline; the window is clamped so scan volume stays bounded no
// matter what the caller asks for. On deadline the query is stopped
// server-side (StopQuery).

import (
	"context"
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs"
	cwltypes "github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs/types"

	"github.com/rladnwls122/skills_dashboard-wooj/backend/internal/config"
)

type InsightsRow map[string]string

type InsightsResult struct {
	Rows           []InsightsRow
	BytesScanned   int64
	RecordsMatched int64
	WindowLabel    string
}

type InsightsParams struct {
	LogGroup string
	Query    string
	WindowMs int64
	// Explicit bounds from the page's shared window. Given these, the query
	// covers exactly the span every other panel covers.
	StartMs  *int64
	EndMs    *int64
	Deadline time.Duration
	// Which region's Logs endpoint holds this group. A CLOUDFRONT-scope WAF log
	// group only exists in us-east-1 — querying it from the workload region
	// returns ResourceNotFoundException, which reads as "logging is off".
	Region string
}

// RunInsightsQuery queues behind a small semaphore — Logs Insights allows few
// concurrent queries per account and each one costs money.
func (a *AWS) RunInsightsQuery(ctx context.Context, p InsightsParams) (InsightsResult, error) {
	deadline := p.Deadline
	if deadline == 0 {
		deadline = config.InsightsLimits.QueryDeadline
	}
	explicit := p.StartMs != nil && p.EndMs != nil
	var windowMs int64
	if explicit {
		windowMs = *p.EndMs - *p.StartMs
	} else if p.WindowMs > 0 {
		windowMs = p.WindowMs
	} else {
		windowMs = config.InsightsLimits.DefaultWindow.Milliseconds()
	}
	// The cap applies either way: it is what bounds bytes scanned.
	if max := config.InsightsLimits.MaxWindow.Milliseconds(); windowMs > max {
		windowMs = max
	}
	endMs := time.Now().UnixMilli()
	if explicit {
		endMs = *p.EndMs
	}
	endSec := endMs / 1000
	startSec := endSec - windowMs/1000
	windowLabel := fmt.Sprintf("%dm", (windowMs+30_000)/60_000)

	region := p.Region
	if region == "" {
		region = a.Settings.Region()
	}
	client, err := a.logs(ctx, region)
	if err != nil {
		return InsightsResult{}, err
	}

	select {
	case a.insightsSem <- struct{}{}:
	case <-ctx.Done():
		return InsightsResult{}, ctx.Err()
	}
	defer func() { <-a.insightsSem }()

	started, err := client.StartQuery(ctx, &cloudwatchlogs.StartQueryInput{
		LogGroupName: aws.String(p.LogGroup),
		StartTime:    aws.Int64(startSec),
		EndTime:      aws.Int64(endSec),
		QueryString:  aws.String(p.Query),
	})
	if err != nil {
		return InsightsResult{}, err
	}
	if started.QueryId == nil {
		return InsightsResult{}, fmt.Errorf("StartQuery failed (no queryId)")
	}
	queryID := started.QueryId

	stop := time.Now().Add(deadline)
	for {
		select {
		case <-time.After(700 * time.Millisecond):
		case <-ctx.Done():
			return InsightsResult{}, ctx.Err()
		}
		res, err := client.GetQueryResults(ctx, &cloudwatchlogs.GetQueryResultsInput{QueryId: queryID})
		if err != nil {
			return InsightsResult{}, err
		}
		switch res.Status {
		case cwltypes.QueryStatusComplete:
			rows := make([]InsightsRow, 0, len(res.Results))
			for _, fields := range res.Results {
				row := InsightsRow{}
				for _, f := range fields {
					if f.Field != nil && f.Value != nil {
						row[*f.Field] = *f.Value
					}
				}
				rows = append(rows, row)
			}
			out := InsightsResult{Rows: rows, WindowLabel: windowLabel}
			if res.Statistics != nil {
				out.BytesScanned = int64(res.Statistics.BytesScanned)
				out.RecordsMatched = int64(res.Statistics.RecordsMatched)
			}
			return out, nil
		case cwltypes.QueryStatusFailed, cwltypes.QueryStatusCancelled, cwltypes.QueryStatusTimeout:
			return InsightsResult{}, fmt.Errorf("Logs Insights query %s", res.Status)
		}
		if time.Now().After(stop) {
			// best effort — the deadline error is the one that matters
			_, _ = client.StopQuery(ctx, &cloudwatchlogs.StopQueryInput{QueryId: queryID})
			return InsightsResult{}, fmt.Errorf("Logs Insights 쿼리 데드라인 초과 (%.0fs)", deadline.Seconds())
		}
	}
}

func FmtBytes(n int64) string {
	switch {
	case n >= 1024*1024*1024:
		return fmt.Sprintf("%.2fGB", float64(n)/1024/1024/1024)
	case n >= 1024*1024:
		return fmt.Sprintf("%.2fMB", float64(n)/1024/1024)
	case n >= 1024:
		return fmt.Sprintf("%.1fKB", float64(n)/1024)
	default:
		return fmt.Sprintf("%dB", n)
	}
}

File: backend/awsx/insights.ts
// CloudWatch Logs Insights runner. One query with a hard deadline; the window is
// clamped so scan volume stays bounded no matter what the caller asks for. On
// deadline the query is stopped server-side (StopQuery).

import {
  GetQueryResultsCommand,
  StartQueryCommand,
  StopQueryCommand,
} from "@aws-sdk/client-cloudwatch-logs";
import { setTimeout as sleep } from "node:timers/promises";

import { INSIGHTS_LIMITS } from "../config/thresholds.ts";
import type { AWS } from "./clients.ts";

export type InsightsRow = Record<string, string>;

export interface InsightsResult {
  rows: InsightsRow[];
  bytesScanned: number;
  recordsMatched: number;
  windowLabel: string;
}

export interface InsightsParams {
  /**
   * One log group, or several separated by commas — ECS task definitions
   * usually give each service its own awslogs group (/ecs/user, /ecs/product,
   * /ecs/stress), and one query over all of them is what the panels want.
   */
  logGroup: string;
  query: string;
  windowMs?: number;
  /**
   * Explicit bounds from the page's shared window. Given these, the query covers
   * exactly the span every other panel covers.
   */
  startMs?: number;
  endMs?: number;
  deadlineMs?: number;
  /**
   * Which region's Logs endpoint holds this group. A CLOUDFRONT-scope WAF log
   * group only exists in us-east-1 — querying it from the workload region
   * returns ResourceNotFoundException, which reads as "logging is off".
   */
  region?: string;
}

/**
 * Queues behind a small semaphore — Logs Insights allows few concurrent queries
 * per account and each one costs money.
 */
export async function runInsightsQuery(a: AWS, p: InsightsParams): Promise<InsightsResult> {
  const deadlineMs = p.deadlineMs ?? INSIGHTS_LIMITS.queryDeadlineMs;
  const explicit = p.startMs !== undefined && p.endMs !== undefined;

  let windowMs: number;
  if (explicit) windowMs = p.endMs! - p.startMs!;
  else if (p.windowMs && p.windowMs > 0) windowMs = p.windowMs;
  else windowMs = INSIGHTS_LIMITS.defaultWindowMs;
  // The cap applies either way: it is what bounds bytes scanned.
  windowMs = Math.min(windowMs, INSIGHTS_LIMITS.maxWindowMs);

  const endMs = explicit ? p.endMs! : Date.now();
  const endSec = Math.floor(endMs / 1000);
  const startSec = endSec - Math.floor(windowMs / 1000);
  const windowLabel = `${Math.floor((windowMs + 30_000) / 60_000)}m`;

  const client = a.logs(p.region || a.settings.region());

  return a.insightsSem.run(async () => {
    const groups = splitLogGroups(p.logGroup);
    const started = await client.send(
      new StartQueryCommand({
        // StartQuery takes either one name or a list, never both.
        ...(groups.length > 1 ? { logGroupNames: groups } : { logGroupName: p.logGroup }),
        startTime: startSec,
        endTime: endSec,
        queryString: p.query,
      }),
    );
    const queryId = started.queryId;
    if (!queryId) throw new Error("StartQuery failed (no queryId)");

    const stopAt = Date.now() + deadlineMs;
    for (;;) {
      await sleep(700);
      const res = await client.send(new GetQueryResultsCommand({ queryId }));

      if (res.status === "Complete") {
        const rows: InsightsRow[] = (res.results ?? []).map((fields) => {
          const row: InsightsRow = {};
          for (const f of fields) {
            if (f.field && f.value !== undefined) row[f.field] = f.value;
          }
          return row;
        });
        return {
          rows,
          bytesScanned: res.statistics?.bytesScanned ?? 0,
          recordsMatched: res.statistics?.recordsMatched ?? 0,
          windowLabel,
        };
      }
      if (res.status === "Failed" || res.status === "Cancelled" || res.status === "Timeout") {
        throw new Error(`Logs Insights query ${res.status}`);
      }
      if (Date.now() > stopAt) {
        // Best effort — the deadline error is the one that matters.
        try {
          await client.send(new StopQueryCommand({ queryId }));
        } catch {
          // ignored
        }
        throw new Error(`Logs Insights 쿼리 데드라인 초과 (${Math.round(deadlineMs / 1000)}s)`);
      }
    }
  });
}

/**
 * Turns the setting's comma-separated value into the list StartQuery takes.
 * A single name comes back as a one-element list.
 */
export function splitLogGroups(raw: string): string[] {
  return raw
    .split(",")
    .map((g) => g.trim())
    .filter((g) => g !== "");
}

export function fmtBytes(n: number): string {
  if (n >= 1024 * 1024 * 1024) return `${(n / 1024 / 1024 / 1024).toFixed(2)}GB`;
  if (n >= 1024 * 1024) return `${(n / 1024 / 1024).toFixed(2)}MB`;
  if (n >= 1024) return `${(n / 1024).toFixed(1)}KB`;
  return `${n}B`;
}
